[Bug][Master] send ACK event timeout

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.


### What happened

In the cluster, a large number of workflows are executed using a serial waiting strategy. It was found that some tasks have been completed, but the workflow has not ended. Checking the logs,  the following logs were found:

At `17:12:43.517`, the master sent the execution result ACK of task `204531` to the worker, but the worker responded with a timeout.
```
[WARN] 2023-12-07 17:12:43.517 +0800 org.apache.dolphinscheduler.extract.base.future.ResponseFuture:[78] - [WorkflowInstance-184914][TaskInstance-204531] - Wait response in 3000/ms timeout, request id 83365
[ERROR] 2023-12-07 17:12:43.517 +0800 org.apache.dolphinscheduler.server.master.processor.queue.TaskExecuteRunnable:[64] - [WorkflowInstance-184914][TaskInstance-204531] - Handle task event error, this event will be removed, event: TaskEvent(taskInstanceId=204531, workerAddress=dolphinscheduler-worker-0.dolphinscheduler-worker-headless:1234, state=TaskExecutionStatus{code=7, desc='success'}, startTime=Thu Dec 07 17:11:47 CST 2023, endTime=Thu Dec 07 17:12:39 CST 2023, executePath=/tmp/dolphinscheduler/exec/process/default/10579449480160/11310276585441_19/184914/204531, logPath=/opt/dolphinscheduler/logs/20231207/11310276585441/19/184914/204531.log, processId=0, appIds=null, event=RESULT, varPool=[], cacheTaskInstanceId=0, processInstanceId=184914)
org.apache.dolphinscheduler.server.master.event.TaskEventHandleError: Handle task result event error, save taskInstance to db error
 at org.apache.dolphinscheduler.server.master.event.TaskResultEventHandler.handleTaskEvent(TaskResultEventHandler.java:105)
 at org.apache.dolphinscheduler.server.master.processor.queue.TaskExecuteRunnable.run(TaskExecuteRunnable.java:56)
 at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
 at java.util.concurrent.FutureTask.run(FutureTask.java:266)
 at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
 at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
 at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.reflect.UndeclaredThrowableException: null
 at com.sun.proxy.$Proxy246.handleTaskInstanceExecutionFinishEventAck(Unknown Source)
 at org.apache.dolphinscheduler.server.master.event.TaskResultEventHandler.sendAckToWorker(TaskResultEventHandler.java:121)
 at org.apache.dolphinscheduler.server.master.event.TaskResultEventHandler.handleTaskEvent(TaskResultEventHandler.java:102)
 ... 6 common frames omitted
Caused by: org.apache.dolphinscheduler.extract.base.exception.RemotingTimeoutException: wait response on the channel Host(ip=dolphinscheduler-worker-0.dolphinscheduler-worker-headless, port=1234) timeout 3000
 at org.apache.dolphinscheduler.extract.base.NettyRemotingClient.sendSync(NettyRemotingClient.java:215)
 at org.apache.dolphinscheduler.extract.base.client.SyncClientMethodInvoker.invoke(SyncClientMethodInvoker.java:46)
 at org.apache.dolphinscheduler.extract.base.client.ClientInvocationHandler.invoke(ClientInvocationHandler.java:55)
 ... 9 common frames omitted

```

At `17:13:02.140`, the following log was found on the worker, indicating that the worker had received the ACK of task `204531`, but there was a delay in processing. The worker system load was relatively high at the time. 
```
[INFO] 2023-12-07 17:13:02.140 +0800 o.a.d.server.worker.runner.listener.TaskInstanceExecutionFinishEventAckListenFunction:[44] - [WorkflowInstance-0][TaskInstance-204531] - Receive TaskInstanceExecutionFinishEventAck: TaskInstanceExecutionFinishEventAck(taskInstanceId=204531, success=true)
```

Checking the code, it was found that an exception occurred when the master sent `sendAckToWorker(taskEvent)`, causing the subsequent process to not be handled, leading to the workflow not ending. However, the worker actually received the ACK, but there was a delay in processing, so the worker will not retry sending the ACK.
https://github.com/apache/dolphinscheduler/blob/ce11674668cf26747f34c3474763c32e1ba5b6d1/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/event/TaskResultEventHandler.java#L90-L106




### What you expected to happen

Everything is OK.

### How to reproduce

...

### Anything else

_No response_

### Version

dev

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/README.md
+++ b/README.md
@@ -3,7 +3,7 @@
 [![License](https://img.shields.io/badge/license-Apache%202-4EB1BA.svg)](https://www.apache.org/licenses/LICENSE-2.0.html)
 ![codecov](https://codecov.io/gh/apache/dolphinscheduler/branch/dev/graph/badge.svg)
 [![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=apache-dolphinscheduler&metric=alert_status)](https://sonarcloud.io/dashboard?id=apache-dolphinscheduler)
-[![Twitter Follow](https://img.shields.io/twitter/follow/dolphinschedule.svg?style=social&label=Follow)](https://twitter.com/dolphinschedule)
+[![Twitter Follow](https://img.shields.io/twitter/follow/dolphinschedule.svg?style=social&label=Follow)](https://twitter.com/dolphinschedule) <!-- markdown-link-check-disable-line -->
 [![Slack Status](https://img.shields.io/badge/slack-join_chat-white.svg?logo=slack&style=social)](https://s.apache.org/dolphinscheduler-slack)
 [![CN doc](https://img.shields.io/badge/文档-中文版-blue.svg)](README_zh_CN.md)
 
@@ -65,7 +65,7 @@ find the good first issue in [here](https://github.com/apache/dolphinscheduler/c
 Welcome to join the Apache DolphinScheduler community by:
 
 - Join the [DolphinScheduler Slack](https://s.apache.org/dolphinscheduler-slack) to keep in touch with the community
-- Follow the [DolphinScheduler Twitter](https://twitter.com/dolphinschedule) and get the latest news
+- Follow the [DolphinScheduler Twitter](https://twitter.com/dolphinschedule) and get the latest news <!-- markdown-link-check-disable-line -->
 - Subscribe DolphinScheduler mail list, users@dolphinscheduler.apache.org for user and dev@dolphinscheduler.apache.org for developer
 
 # Landscapes
@@ -75,5 +75,4 @@ Welcome to join the Apache DolphinScheduler community by:
 <img src="https://landscape.cncf.io/images/left-logo.svg" width="150"/>&nbsp;&nbsp;<img src="https://landscape.cncf.io/images/right-logo.svg" width="200"/>
 <br/><br/>
 DolphinScheduler enriches the <a href="https://landscape.cncf.io/?landscape=observability-and-analysis&license=apache-license-2-0">CNCF CLOUD NATIVE Landscape.</a >
-
 </p >

--- a/README_zh_CN.md
+++ b/README_zh_CN.md
@@ -3,7 +3,7 @@
 [![License](https://img.shields.io/badge/license-Apache%202-4EB1BA.svg)](https://www.apache.org/licenses/LICENSE-2.0.html)
 [![codecov](https://codecov.io/gh/apache/dolphinscheduler/branch/dev/graph/badge.svg)]()
 [![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=apache-dolphinscheduler&metric=alert_status)](https://sonarcloud.io/dashboard?id=apache-dolphinscheduler)
-[![Twitter Follow](https://img.shields.io/twitter/follow/dolphinschedule.svg?style=social&label=Follow)](https://twitter.com/dolphinschedule)
+[![Twitter Follow](https://img.shields.io/twitter/follow/dolphinschedule.svg?style=social&label=Follow)](https://twitter.com/dolphinschedule) <!-- markdown-link-check-disable-line -->
 [![Slack Status](https://img.shields.io/badge/slack-join_chat-white.svg?logo=slack&style=social)](https://s.apache.org/dolphinscheduler-slack)
 [![EN doc](https://img.shields.io/badge/document-English-blue.svg)](README.md)
 
@@ -61,7 +61,7 @@ DolphinScheduler 的主要特性如下：
 欢迎通过以方式加入社区：
 
 - 加入 [DolphinScheduler Slack](https://s.apache.org/dolphinscheduler-slack)
-- 关注 [DolphinScheduler Twitter](https://twitter.com/dolphinschedule) 来获取最新消息
+- 关注 [DolphinScheduler Twitter](https://twitter.com/dolphinschedule) 来获取最新消息 <!-- markdown-link-check-disable-line -->
 - 订阅 DolphinScheduler 邮件列表, 用户订阅 users@dolphinscheduler.apache.org 开发者请订阅 dev@dolphinscheduler.apache.org
 
 # Landscapes

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/event/TaskResultEventHandler.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/event/TaskResultEventHandler.java
@@ -128,7 +128,7 @@ public class TaskResultEventHandler implements TaskEventHandler {
                     TaskInstanceExecutionFinishEventAck.success(taskEvent.getTaskInstanceId()));
         } catch (Exception e) {
             // master ignore the exception, worker will retry to send this TaskEventType.RESULT event again.
-            log.error("send ack to worker error, taskInstanceId: {}", taskEvent.getTaskInstanceId(), e);
+            log.warn("send ack to worker error, taskInstanceId: {}", taskEvent.getTaskInstanceId(), e);
         }
     }
 

--- a/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/event/TaskResultEventHandler.java
+++ b/dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/event/TaskResultEventHandler.java
@@ -35,10 +35,13 @@ import org.apache.dolphinscheduler.service.process.ProcessService;
 
 import java.util.Optional;
 
+import lombok.extern.slf4j.Slf4j;
+
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
 
 @Component
+@Slf4j
 public class TaskResultEventHandler implements TaskEventHandler {
 
     @Autowired
@@ -99,11 +102,13 @@ public class TaskResultEventHandler implements TaskEventHandler {
             taskInstance.setVarPool(taskEvent.getVarPool());
             processService.changeOutParam(taskInstance);
             taskInstanceDao.updateById(taskInstance);
-            sendAckToWorker(taskEvent);
         } catch (Exception ex) {
             TaskInstanceUtils.copyTaskInstance(oldTaskInstance, taskInstance);
             throw new TaskEventHandleError("Handle task result event error, save taskInstance to db error", ex);
         }
+
+        sendAckToWorker(taskEvent);
+
         TaskStateEvent stateEvent = TaskStateEvent.builder()
                 .processInstanceId(taskEvent.getProcessInstanceId())
                 .taskInstanceId(taskEvent.getTaskInstanceId())
@@ -115,11 +120,16 @@ public class TaskResultEventHandler implements TaskEventHandler {
     }
 
     public void sendAckToWorker(TaskEvent taskEvent) {
-        ITaskInstanceExecutionEventAckListener instanceExecutionEventAckListener =
-                SingletonJdkDynamicRpcClientProxyFactory
-                        .getProxyClient(taskEvent.getWorkerAddress(), ITaskInstanceExecutionEventAckListener.class);
-        instanceExecutionEventAckListener.handleTaskInstanceExecutionFinishEventAck(
-                TaskInstanceExecutionFinishEventAck.success(taskEvent.getTaskInstanceId()));
+        try {
+            ITaskInstanceExecutionEventAckListener instanceExecutionEventAckListener =
+                    SingletonJdkDynamicRpcClientProxyFactory
+                            .getProxyClient(taskEvent.getWorkerAddress(), ITaskInstanceExecutionEventAckListener.class);
+            instanceExecutionEventAckListener.handleTaskInstanceExecutionFinishEventAck(
+                    TaskInstanceExecutionFinishEventAck.success(taskEvent.getTaskInstanceId()));
+        } catch (Exception e) {
+            // master ignore the exception, worker will retry to send this TaskEventType.RESULT event again.
+            log.error("send ack to worker error, taskInstanceId: {}", taskEvent.getTaskInstanceId(), e);
+        }
     }
 
     @Override

